lib/mime.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+const { parse } = internalBinding('mime_wrap');
+const {
+  Apply,
+  MapIteratorPrototypeNext,
+  RegexpPrototypeTest,
+  StringPrototypeReplace,
+  SymbolReplace,
+  SymbolIterator,
+  TypeError,
+  SafeMap: Map,
+  SafeWeakMap: WeakMap,
+} = require('internal/safe_globals');
+
+const DATA = new WeakMap();
+const $ = (self, guard) => {
+  if (!DATA.has(self)) {
+    throw new TypeError('Invalid receiver');
+  }
+  const data = DATA.get(self);
+  if (guard !== data.guard) {
+    throw new TypeError('Invalid receiver');
+  }
+  return data.value;
+};
+const NotHTTPTokenCodePoint = /[^\u0021\u0023-\u0027\u002A\u002B\u002D\u002E\u005E\u005F\u0060\u007C\u007EA-Za-z0-9]/g;
+NotHTTPTokenCodePoint[SymbolReplace] = null;
+const encode = (value) => {
+  NotHTTPTokenCodePoint.lastIndex = 0;
+  const encode = Apply(RegexpPrototypeTest, NotHTTPTokenCodePoint, [value]);
+  if (!encode) return value;
+  const escaped = Apply(StringPrototypeReplace, value,
+                        [/["\\]/g, '\\$&']);
+  const ret = `"${escaped}"`;
+  return ret;
+};
+class MIMEParams {
+  constructor() {
+    DATA.set(this, {
+      value: new Map(),
+      guard: MIMEParams
+    });
```

Comment:
i don't want it exposed so this is just how it is until we get private fields. I'm not going to look at symbols because they are a public part of APIs.

Reply:
Hmmm... I'm with @addaleax on this. I'd use private fields also if they were available now but this just feels like an overly complicated hack that is inconsistent with how we do this in other parts of core. 

A non enumerable symbol should be private *enough*.

Reply:
@jasnell I'd disagree and do believe that symbols also add noise and complexity. Might just be me though. Matching how private fields work makes the transition to them smooth and alleviates any concerns about people using internals.

Reply:
I'm onboard with using WeakMaps but the guard thing seems a bit off.  Why don't you use separate WeakMaps for MIME and MIMEParams?

Reply:
bnoordhuis seems fine, I'll probably change the `$()` helper to accept a parameter for which map to check in instead still though. `guard` was to make sure it was coming from the right place. We won't need that with 2 WeakMaps but I'd prefer the receiver error be in a single place :).

Reply:
fixed